Error fetching file: Unknown paint type: PATTERN

**Describe the bug**
I get 
```
{
  "fileKey": "aH0X1gy9OicGnOnG6fMpdK"
}
```
`Error fetching file: Unknown paint type: PATTERN`

**Software Versions**

- Figma Developer MCP: npx, 0.4.3
- Node.js: v18.20.0
- NPM: `10.9.0`
- Operating System: Ubuntu 24
- Client: e.g. Cursor
- Client Version: latest

**To Reproduce**
Steps to reproduce the behavior:

Asked the agent to `implement @https://www.figma.com/design/randomseed/project?node-id=336-3901&m=dev in html, css and vanilla js as index in ./frontend`

**Expected behavior**
The actual output made, the design is a very sample landing page with a title, a button, an image.


**Server Configuration**
Provide your MCP JSON configuration, if applicable. E.g.

```
 {
  "mcpServers": {
    "Framelink Figma MCP": {
      "command": "npx",
      "args": ["-y", "figma-developer-mcp", "--figma-api-key=mykey", "--stdio"]
    }
  }
}
```

**Command Line Logs**
If you're running the MCP locally on the command line, include all the logs for those like so:
```
{
  "mcpServers": {
    "Framelink Figma MCP": {
      "command": "npx",
      "args": ["-y", "figma-developer-mcp", "--figma-api-key=figd_wdLGahyDKnMGaUToypdg6SejbDtALChNXt0Ztfj3", "--stdio"]
    }
  }
}
```


**MCP Logs**


```
2025-07-08 20:05:57.030 [info] user-Framelink Figma MCP: Handling CallTool action for tool 'get_figma_data'
2025-07-08 20:05:57.031 [info] user-Framelink Figma MCP: Calling tool 'get_figma_data' with toolCallId: fileid
2025-07-08 20:05:57.038 [error] user-Framelink Figma MCP: [INFO] Fetching all layers of node 336-3901 from file fileid

2025-07-08 20:05:57.038 [error] user-Framelink Figma MCP: [INFO] Fetching all layers of node 336-3901 from file fileid

2025-07-08 20:05:57.039 [error] user-Framelink Figma MCP: [INFO] Calling https://api.figma.com/v1/files/fileid/nodes?ids=336-3901
[INFO] Using Personal Access Token for authentication

2025-07-08 20:05:57.039 [error] user-Framelink Figma MCP: [INFO] Calling https://api.figma.com/v1/files/fileid/nodes?ids=336-3901
[INFO] Using Personal Access Token for authentication

2025-07-08 20:05:58.883 [error] user-Framelink Figma MCP: [INFO] Got response from getNode, now parsing.

2025-07-08 20:05:58.883 [error] user-Framelink Figma MCP: [INFO] Got response from getNode, now parsing.

2025-07-08 20:05:58.885 [error] user-Framelink Figma MCP: [ERROR] Error fetching file fileid: Unknown paint type: PATTERN

2025-07-08 20:05:58.885 [error] user-Framelink Figma MCP: [ERROR] Error fetching file fileid: Unknown paint type: PATTERN

2025-07-08 20:05:58.887 [info] user-Framelink Figma MCP: Successfully called tool 'get_figma_data'

```

--- a/.changeset/many-sides-wash.md
+++ b/.changeset/many-sides-wash.md
@@ -1,0 +1,5 @@
+---
+"figma-developer-mcp": minor
+---
+
+Add support for pattern fills in Figma.

--- a/src/transformers/style.ts
+++ b/src/transformers/style.ts
@@ -251,64 +251,6 @@ export function buildSimplifiedStrokes(
 }
 
 /**
- * Convert a Figma PatternPaint to a CSS-like pattern fill.
- *
- * Ignores `tileType` and `spacing` from the Figma API currently as there's
- * no great way to translate them to CSS.
- *
- * @param raw - The Figma PatternPaint to convert
- * @returns The converted pattern SimplifiedFill
- */
-function parsePatternPaint(
-  raw: Extract<Paint, { type: "PATTERN" }>,
-): Extract<SimplifiedFill, { type: "PATTERN" }> {
-  /**
-   * The only CSS-like repeat value supported by Figma is repeat.
-   *
-   * They also have hexagonal horizontal and vertical repeats, but
-   * those aren't easy to pull off in CSS, so we just use repeat.
-   */
-  let backgroundRepeat = "repeat";
-
-  let horizontal = "left";
-  switch (raw.horizontalAlignment) {
-    case "START":
-      horizontal = "left";
-      break;
-    case "CENTER":
-      horizontal = "center";
-      break;
-    case "END":
-      horizontal = "right";
-      break;
-  }
-
-  let vertical = "top";
-  switch (raw.verticalAlignment) {
-    case "START":
-      vertical = "top";
-      break;
-    case "CENTER":
-      vertical = "center";
-      break;
-    case "END":
-      vertical = "bottom";
-      break;
-  }
-
-  return {
-    type: raw.type,
-    patternSource: {
-      type: "IMAGE-PNG",
-      nodeId: raw.sourceNodeId,
-    },
-    backgroundRepeat,
-    backgroundSize: `${Math.round(raw.scalingFactor * 100)}%`,
-    backgroundPosition: `${horizontal} ${vertical}`,
-  };
-}
-
-/**
  * Convert a Figma paint (solid, image, gradient) to a SimplifiedFill
  * @param raw - The Figma paint to convert
  * @param hasChildren - Whether the node has children (determines CSS properties)
@@ -381,6 +323,64 @@ export function parsePaint(raw: Paint, hasChildren: boolean = false): Simplified
   } else {
     throw new Error(`Unknown paint type: ${raw.type}`);
   }
+}
+
+/**
+ * Convert a Figma PatternPaint to a CSS-like pattern fill.
+ *
+ * Ignores `tileType` and `spacing` from the Figma API currently as there's
+ * no great way to translate them to CSS.
+ *
+ * @param raw - The Figma PatternPaint to convert
+ * @returns The converted pattern SimplifiedFill
+ */
+function parsePatternPaint(
+  raw: Extract<Paint, { type: "PATTERN" }>,
+): Extract<SimplifiedFill, { type: "PATTERN" }> {
+  /**
+   * The only CSS-like repeat value supported by Figma is repeat.
+   *
+   * They also have hexagonal horizontal and vertical repeats, but
+   * those aren't easy to pull off in CSS, so we just use repeat.
+   */
+  let backgroundRepeat = "repeat";
+
+  let horizontal = "left";
+  switch (raw.horizontalAlignment) {
+    case "START":
+      horizontal = "left";
+      break;
+    case "CENTER":
+      horizontal = "center";
+      break;
+    case "END":
+      horizontal = "right";
+      break;
+  }
+
+  let vertical = "top";
+  switch (raw.verticalAlignment) {
+    case "START":
+      vertical = "top";
+      break;
+    case "CENTER":
+      vertical = "center";
+      break;
+    case "END":
+      vertical = "bottom";
+      break;
+  }
+
+  return {
+    type: raw.type,
+    patternSource: {
+      type: "IMAGE-PNG",
+      nodeId: raw.sourceNodeId,
+    },
+    backgroundRepeat,
+    backgroundSize: `${Math.round(raw.scalingFactor * 100)}%`,
+    backgroundPosition: `${horizontal} ${vertical}`,
+  };
 }
 
 /**

--- a/src/transformers/style.ts
+++ b/src/transformers/style.ts
@@ -80,6 +80,20 @@ export type SimplifiedGradientFill = {
 
 export type SimplifiedPatternFill = {
   type: "PATTERN";
+  patternSource: {
+    /**
+     * Hardcode to expect PNG for now, consider SVG detection in the future.
+     *
+     * SVG detection is a bit challenging because the nodeId in question isn't
+     * guaranteed to be included in the response we're working with. No guaranteed
+     * way to look into it and see if it's only composed of vector shapes.
+     */
+    type: "IMAGE-PNG";
+    nodeId: string;
+  };
+  backgroundRepeat: string;
+  backgroundSize: string;
+  backgroundPosition: string;
 };
 
 export type SimplifiedFill =
@@ -235,6 +249,65 @@ export function buildSimplifiedStrokes(
 
   return strokes;
 }
+
+/**
+ * Convert a Figma PatternPaint to a CSS-like pattern fill.
+ *
+ * Ignores `tileType` and `spacing` from the Figma API currently as there's
+ * no great way to translate them to CSS.
+ *
+ * @param raw - The Figma PatternPaint to convert
+ * @returns The converted pattern SimplifiedFill
+ */
+function parsePatternPaint(
+  raw: Extract<Paint, { type: "PATTERN" }>,
+): Extract<SimplifiedFill, { type: "PATTERN" }> {
+  /**
+   * The only CSS-like repeat value supported by Figma is repeat.
+   *
+   * They also have hexagonal horizontal and vertical repeats, but
+   * those aren't easy to pull off in CSS, so we just use repeat.
+   */
+  let backgroundRepeat = "repeat";
+
+  let horizontal = "left";
+  switch (raw.horizontalAlignment) {
+    case "START":
+      horizontal = "left";
+      break;
+    case "CENTER":
+      horizontal = "center";
+      break;
+    case "END":
+      horizontal = "right";
+      break;
+  }
+
+  let vertical = "top";
+  switch (raw.verticalAlignment) {
+    case "START":
+      vertical = "top";
+      break;
+    case "CENTER":
+      vertical = "center";
+      break;
+    case "END":
+      vertical = "bottom";
+      break;
+  }
+
+  return {
+    type: raw.type,
+    patternSource: {
+      type: "IMAGE-PNG",
+      nodeId: raw.sourceNodeId,
+    },
+    backgroundRepeat,
+    backgroundSize: `${Math.round(raw.scalingFactor * 100)}%`,
+    backgroundPosition: `${horizontal} ${vertical}`,
+  };
+}
+
 /**
  * Convert a Figma paint (solid, image, gradient) to a SimplifiedFill
  * @param raw - The Figma paint to convert
@@ -287,9 +360,7 @@ export function parsePaint(raw: Paint, hasChildren: boolean = false): Simplified
       return formatRGBAColor(raw.color!, opacity);
     }
   } else if (raw.type === "PATTERN") {
-    return {
-      type: raw.type,
-    };
+    return parsePatternPaint(raw);
   } else if (
     ["GRADIENT_LINEAR", "GRADIENT_RADIAL", "GRADIENT_ANGULAR", "GRADIENT_DIAMOND"].includes(
       raw.type,

